File: chronoid-ts/src/SnowflakeGenerator.ts
import { AsyncSnowflakeGenerator } from './AsyncSnowflakeGenerator.js';
import { SyncSnowflakeGenerator } from './SyncSnowflakeGenerator.js';
import { AsyncExhaustionStrategy, SyncExhaustionStrategy } from './types.js';
import { InvalidNodeId, InvalidWorkerId } from './errors.js';

/**
 * Static factory responsible for constructing and validating generator instances.
 */
export class SnowflakeGenerator {
  public static create(
    base_year: number,
    node_id: number,
    worker_id: number,
    strategy: AsyncExhaustionStrategy
  ): AsyncSnowflakeGenerator {
    SnowflakeGenerator.validateConfig(base_year, node_id, worker_id);
    return new AsyncSnowflakeGenerator(base_year, node_id, worker_id, strategy);
  }

  public static create_sync(
    base_year: number,
    node_id: number,
    worker_id: number,
    strategy: SyncExhaustionStrategy
  ): SyncSnowflakeGenerator {
    SnowflakeGenerator.validateConfig(base_year, node_id, worker_id);
    return new SyncSnowflakeGenerator(base_year, node_id, worker_id, strategy);
  }

  private static validateConfig(base_year: number, node_id: number, worker_id: number): void {
    if (base_year < 1900 || base_year > 2200) {
      throw new Error(`base_year must be between 1900 and 2200. Got: ${base_year}`);
    }
    if (node_id < 0 || node_id > 15) {
      throw new InvalidNodeId(`node_id must be between 0 and 15. Got: ${node_id}`);
    }
    if (worker_id < 0 || worker_id > 15) {
      throw new InvalidWorkerId(`worker_id must be between 0 and 15. Got: ${worker_id}`);
    }
  }
}

File: chronoid-java/src/main/java/io/github/datakore/chronoid/SnowflakeGenerator.java
package io.github.datakore.chronoid;

public class SnowflakeGenerator {
    private SnowflakeGenerator() {}

    public static AsyncSnowflakeGenerator create(
        int baseYear,
        int nodeId,
        int workerId,
        AsyncExhaustionStrategy strategy
    ) {
        validateConfig(baseYear, nodeId, workerId);
        return new AsyncSnowflakeGenerator(baseYear, nodeId, workerId, strategy);
    }

    public static SyncSnowflakeGenerator createSync(
        int baseYear,
        int nodeId,
        int workerId,
        SyncExhaustionStrategy strategy
    ) {
        validateConfig(baseYear, nodeId, workerId);
        return new SyncSnowflakeGenerator(baseYear, nodeId, workerId, strategy);
    }

    private static void validateConfig(int baseYear, int nodeId, int workerId) {
        if (baseYear < 1900 || baseYear > 2200) {
            throw new ChronoidException.InvalidId("baseYear must be between 1900 and 2200. Got: " + baseYear);
        }
        if (nodeId < 0 || nodeId > 15) {
            throw new ChronoidException.InvalidId("nodeId must be between 0 and 15. Got: " + nodeId);
        }
        if (workerId < 0 || workerId > 15) {
            throw new ChronoidException.InvalidId("workerId must be between 0 and 15. Got: " + workerId);
        }
    }
}
